c10/util/TypeCast.h

```diff
@@ -103,16 +103,35 @@ C10_HOST_DEVICE inline dest_t static_cast_with_inter_type(src_t src) {
 #define ERROR_UNSUPPORTED_CAST TORCH_CHECK(false, "Unexpected scalar type");
 #endif
 
+template<typename dest_t, typename complex_src_t>
+struct MaybeReal {
+  static C10_HOST_DEVICE inline typename complex_src_t::value_type maybe_real(complex_src_t src) {
+    return src.real();
+  }
+};
+
+template<typename complex_src_t>
+struct MaybeReal<std::complex<float>, complex_src_t> {
+  static C10_HOST_DEVICE inline complex_src_t maybe_real(complex_src_t src) {
+    return src;
+  }
+};
+
+template<typename complex_src_t>
+struct MaybeReal<std::complex<double>, complex_src_t> {
+  static C10_HOST_DEVICE inline complex_src_t maybe_real(complex_src_t src) {
+    return src;
+  }
+};
+
```

Comment:
@zasdfgbnm 

Thanks for adding MaybeReal. I'm surprised that other scalar_t types like `Bool, Half, BFFloat16` can call the `scalar_t::real()` method, but I guess it worked for `std::real(scalar_t)`?

Reply:
@dylanbespalko No, the way it is used makes sure that `complex_src_t` could only be `std::complex<something_t>`.

Reply:
Also, you could really help me out by adding this change to c10/util/TypeCast.h::line48::52. I think it should be:
```C++
template <typename dest_t, typename src_t>
C10_HOST_DEVICE inline dest_t static_cast_with_inter_type(src_t src) {
  return static_cast<dest_t>(
    static_cast<inter_copy_type_t<dest_t>>(MaybeReal<dest_t, src_t>(src)));
}
```

Reply:
Oh, I see. Clever...

Reply:
@dylanbespalko Thanks for the review. That makes great sense. And I will add that.